src/main/java/com/google/firebase/auth/hash/Scrypt.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.auth.hash;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.io.BaseEncoding;
+import com.google.firebase.auth.UserImportHash;
+import java.util.Map;
+
+/**
+ * Represents the Scrypt password hashing algorithm. This is the
+ * <a href="https://github.com/firebase/scrypt">modified Scrypt algorithm</a> used by
+ * Firebase Auth. See {@link StandardScrypt} for the standard Scrypt algorithm. Can be used as an
+ * instance of {@link com.google.firebase.auth.UserImportHash} when importing users.
+ */
+public final class Scrypt extends UserImportHash {
```

Comment:
Can we make this inherit BasicHash and re-use `setRounds()`? On that note, should `BasicHash` be renamed to better reflect what it does?

Reply:
It seems there are 5 types of hash algorithms that we need to support:

1. HMAC hashes
2. MD5, SHA, PBKDF etc
3. Scrypt
4. StandardScrypt
5. Bcrypt

I refer to the 2nd category as basic hashes for the lack of a better name. I think it makes sense to not have Scrypt extend from BasicHash (the allowed integer ranges are different). But I'm certainly open to calling BasicHash something else. Any suggestions?

Reply:
What do you think of making BasicHash a "RepeatableHash" interface? You will have to verify the range in each implementing class, but it's a nicely named interface and would convey some property about the hash functions that adhere to it.

Reply:
Thanks for the suggestion. I renamed `BasicHash` to `RepeatableHash`, and passed range interval as constructor parameters. This enables us to reuse it in `Scrypt` as well.